util/src/MultiSizeBufferPool.cpp

```diff
@@ -0,0 +1,688 @@
+}
+
+MultiSizeBufferPool::SubPool& MultiSizeBufferPool::getSubPool(uint32_t bufferSize) const {
+  auto iter = subpools_.find(bufferSize);
+  ConcordAssert(iter != subpools_.end());  // internal bug or memory corrupted - caller must validate
+  return *iter->second.get();
+}
+
+std::pair<char*, uint32_t> MultiSizeBufferPool::popBuffer(uint32_t requestedBufferSize) {
+  SubPool::AllocationResult result{};
+  uint32_t allocatedBufferSize{};
+  bool allocate{true};
+
+  DEBUG_PRINT(logger_, "pop enter:" << KVLOG(KVLOG_PREFIX));
+  /** 1st stage: repetitive, until reaching the largest subpool:
+   * 1a) Attempt to obtain a chunk from the target subpool that matches bufferSize.
+   * 1b) If such chunk is found - return chunk and exit (success), If not continue to 1c.
+   * 1c) If there are no available chunks in the target subpool, try to allocate one from the next subpool (with a
+   * larger chunk size), repeat 1a->1c. If reached largest subpool, move to stage 2. 2nd stage: 2a) Go back to the
+   * target subpool (which matches bufferSize). 2b) Try to allocate a new chunk and/or wait for chunk to be released by
+   * another consumer. This stage is blocking so a chunk must be found at some time.
+   */
+  auto iter = bufferSizes_.find(requestedBufferSize);
+  ConcordAssertNE(iter, bufferSizes_.end());
+  if (iter == bufferSizes_.end())
+    throwException<std::invalid_argument>(logger_, "Invalid buffer size!", KVLOG(KVLOG_PREFIX, requestedBufferSize));
+
+  do {
+    result = getSubPool(*iter).getChunk(false, allocate);
+    if (result.first) {
+      allocatedBufferSize = *iter;
+      break;
+    }
+    // continue on all errors codes (try next subpool)
+    allocate = false;
+    ++iter;
+  } while (iter != bufferSizes_.end());
+
+  if (result.second == SubPool::AllocationStatus::EXCEED_POOL_MAX_BYTES) {
+    LOG_WARN(logger_,
+             "The pool has reached the maximum pool total bytes allocation threshold (net buffers size "
+             "plus metadata)!"
+                 << KVLOG(KVLOG_PREFIX, config_.maxAllocatedBytes) << stats_);
+  }
+  if (!result.first) {
+    // 2nd stage, if didn't get a buffer: wait infinitely on the initial subpool until a chunk is allocated
+    result = getSubPool(requestedBufferSize).getChunk(true, true);
+    allocatedBufferSize = requestedBufferSize;
+  }
+
+  DEBUG_PRINT(logger_, "pop exit:" << KVLOG(KVLOG_PREFIX));
+  if (result.first) {
+    ConcordAssertGT(allocatedBufferSize, 0);
+    return std::make_pair(chunkToBuffer(result.first), allocatedBufferSize);
+  }
+  return std::make_pair(nullptr, 0);
+}
+
+std::pair<char*, uint32_t> MultiSizeBufferPool::getBufferBySubpoolSelector() {
+  TimeRecorder scoped_timer(*histograms_.time_in_getBufferBySubpoolSelector);
+  DEBUG_PRINT(logger_, "getBufferBySubpoolSelector enter:" << KVLOG(KVLOG_PREFIX));
+  const auto& subpoolConfig{subPoolSelector_->selectSubpool()};
```

Comment:
No, this is completely contrary to my design.
Subpool selector is choosing a buffer size, not a subpool.
the name is confusing...won't call it "buffer size selector" since it's also confusing.
It should be decoupled completely from implementation and shiuldn't know about subpool ot its memebers.